src/noit_metric_tag_search.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2018, Circonus, Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ *       copyright notice, this list of conditions and the following
+ *       disclaimer in the documentation and/or other materials provided
+ *       with the distribution.
+ *     * Neither the name Circonus, Inc. nor the names of its contributors
+ *       may be used to endorse or promote products derived from this
+ *       software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef NOIT_METRIC_TAG_SEARCH_H_
+#define NOIT_METRIC_TAG_SEARCH_H_
+
+/* This implement filtering based on a set of tags where tags are of the form key:value.
+ *
+ * Tag keys and values may contain only characters [a-zA-Z0-9._-].
+ *
+ * The grammar for matching a set of tags supports and/or/not.
+ *
+ * stmt        :=  and(stmt-list)
+ *              |  or(stmt-list)
+ *              |  not(stmt-single)
+ * stmt-list   :=  stmt-single,stmt-list
+ *              |  stmt-single
+ * stmt-single :=  stmt
+ *              |  match
+ * match       :=  partmatch
+ *              |  partmatch : partmatch
+ * partmatch   :=  [a-zA-Z0-9._-*]  // tag character plus wildcard
+ *              |  /<re>/           // re is a regular expression
+ *
+ * example: and(env:prod,or(service:web,team:/^(red|blue)$/),not(special))
```

Comment:
I think the `not` clause will need to be a list

Reply:
How does not apply to a list?  Traditionally not or ! in most languages is a unary operator.

Reply:
yes, I see.. misread the grammar